test/dynamo/test_misc.py

```diff
@@ -828,6 +828,48 @@ def forward(self, arg0_1: "f32[3]", arg1_1: "f32[3]", arg2_1: "f32[3]", arg3_1:
             cleanup_op("mylib::foo")
             del lib
 
+    def test_auto_functionalize_tensorlist(self):
+        try:
+            lib = torch.library.Library("mylib", "FRAGMENT")
+            torch.library.define(
+                "mylib::foo",
+                "(Tensor all_gather_output, SymInt[] all_gather_input_split_sizes, int dim=0, *, Tensor(a!)[] out) -> ()",
+                tags=torch.Tag.pt2_compliant_tag,
+                lib=lib,
+            )
+
+            @torch.library.impl("mylib::foo", "cpu", lib=lib)
+            @torch._dynamo.disable
+            def foo_impl(
+                all_gather_output, all_gather_input_split_sizes, dim=0, *, out
+            ):
+                for o in out:
+                    o.copy_(all_gather_output)
+
+            def f(all_gather_output, all_gather_input_split_sizes, dim, out):
+                torch.ops.mylib.foo(
+                    all_gather_output, all_gather_input_split_sizes, dim, out=out
+                )
+
+            a = torch.ones(4)
+            b = [2, 3]
+            c = 0
+            d = [torch.empty(4) for _ in range(2)]
+            orig_args = (a, b, c, d)
+
+            compiled_args = pytree.tree_map_only(torch.Tensor, torch.clone, orig_args)
+            with ctx():
+                torch.compile(f, backend="aot_eager", fullgraph=True)(
```

Comment:
What happens with backend = inductor? Does it raise an error? If so, what does it look like?

Reply:
the op as I wrote it (borrowed from @yf225) has an out= kwarg, so inductor fails on an assert that it doesn't support out= https://github.com/pytorch/pytorch/blob/main/torch/_inductor/lowering.py#L290

I just tried making the op positional-only, and we fail a different assert here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L5314

I originally figured I would not bother biting off out= handling in inductor, so I would leave inductor alone. But I guess that's an independent issue from mutable list[Tensor] handling in inductor.  Let me know what you think - if you think it's worth getting mutable list[tensor] handling working in inductor, happy to do it as part of this PR.

(my tentative thought at first was that it wasn't worth "fixing" in inductor unless we also do the work of properlying getting list[tensor] support in inductor's reinplacing pass, since we probably never want the functional op to show up at runtime, but maybe they're worth tackling separately)

Reply:
> I just tried making the op positional-only, and we fail a different assert here: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/ir.py#L5314

Can we just make the error message here better? ("You might be here if you have an op that has a mutable List[Tensor], we don't support that yet; file an issue"). 

> if you think it's worth getting mutable list[tensor] handling working in inductor, happy to do it as part of this PR.

Nah seems like too much work for now

Reply:
sg, made the error message in inductor a bit better